### PR TITLE
Generic factory commands

### PR DESCRIPTION
**Pull request by mdellweg**
~This is based on the lazy entity from #71 .~

@daviddavis Have a look at the `pulp file repository show` command at this point of the experiment.

### CODE REVIEW

File: pulpcore/cli/common/generic.py [hunk 1]
```diff
@@ -15,22 +15,218 @@
...
+name_option = click.option(
+    "--name",
+    help="Name of the {entity}",
+    callback=_name_callback,
+    expose_value=False,
+    cls=PulpOption,
+)
+
+version_option = click.option(
+    "--version", help="Version number", type=int, callback=_version_callback, expose_value=False
```
Comment (daviddavis) [inline pulpcore/cli/common/generic.py:101]:
```suggestion
    "--version", help="Repository version number", type=int, callback=_version_callback, expose_value=False
```

File: pulpcore/cli/common/generic.py [hunk 2]
```diff
@@ -15,22 +15,218 @@
...
+        entity_ctx.delete(entity_ctx.pulp_href)
+
+    for option in decorators:
+        # Decorate callback
+        callback = option(callback)
+    return callback
+
+
+def version_command(**kwargs: Any) -> click.Command:
+    """A factory that creates a version command group."""
```
Comment (daviddavis) [inline pulpcore/cli/common/generic.py:186]:
```suggestion
    """A factory that creates a repository version command group."""
```

File: pulpcore/cli/common/generic.py [hunk 3]
```diff
@@ -15,22 +15,218 @@
...
+        entity_ctx: PulpEntityContext = ctx.find_object(PulpEntityContext)
+        entity_ctx.entity = {"name": value}
+    return value
+
+
+def _version_callback(ctx: click.Context, param: click.Parameter, value: int) -> int:
+    if value is not None:
+        entity_ctx: PulpEntityContext = ctx.find_object(PulpEntityContext)
+        repository_ctx: PulpRepositoryContext = ctx.find_object(PulpRepositoryContext)
+        entity_ctx.pulp_href = repository_ctx.entity["versions_href"] + str(value)
```
Comment (gerrod3) [inline pulpcore/cli/common/generic.py:61]: How about we allow the version option to be nullable and have the CLI choose the latest version.
```suggestion
def _version_callback(ctx: click.Context, param: click.Parameter, value: int) -> int:
      entity_ctx: PulpEntityContext = ctx.find_object(PulpEntityContext)
      repository_ctx: PulpRepositoryContext = ctx.find_object(PulpRepositoryContext)
      if value is not None:
          entity_ctx.pulp_href = repository_ctx.entity["versions_href"] + str(value)
      else:
          entity_ctx.pulp_href = repository_ctx.entity["latest_version_href"]
```

**Review by daviddavis — APPROVED** [review]
LGTM :100:

File: pulpcore/cli/common/generic.py (hunk 3 above)
Comment (mdellweg) [reply pulpcore/cli/common/generic.py:61]: That could work, and spare us the embarrassment of trying to do something without having the version specified.

**Pull request merged**